File: app/src/main/java/com/example/foodoutdated/Product.java
package com.example.foodoutdated;

import android.os.Parcel;
import android.os.Parcelable;


import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

public class Product implements Parcelable {
    private String name;
    private Date expiry;        //date format: yyyy-mm-dd
    private String thumbnail;

    // constructor
    public Product() {
        this.name = "";
        this.expiry = new Date();
        this.thumbnail = "";
    }

    public Product(String name, Date expiry, String thumbnail) {
        this.name = name;
        this.expiry = expiry;
        this.thumbnail = thumbnail;
    }


    // getter and setter
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Date getExpiry() {
        return expiry;
    }

    public void setExpiry(Date expiry) {
        this.expiry = expiry;
    }

    public String getThumbnail() {
        return thumbnail;
    }

    public void setThumbnail(String thumbnail) {
        this.thumbnail = thumbnail;
    }


    // Parcelable implement
    protected Product(Parcel in) {
        try {
            name = in.readString();
            thumbnail = in.readString();
            expiry = new SimpleDateFormat(Utils.DATE_PATTERN_VN).parse(in.readString());
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }

    public static final Creator<Product> CREATOR = new Creator<Product>() {
        @Override
        public Product createFromParcel(Parcel in) {
            return new Product(in);
        }

        @Override
        public Product[] newArray(int size) {
            return new Product[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeString(thumbnail);
        dest.writeString(new SimpleDateFormat(Utils.DATE_PATTERN_VN).format(expiry));
    }
}

File: app/src/main/java/com/ninh/foodoutdated/Product.kt
package com.ninh.foodoutdated

import android.os.Parcel
import android.os.Parcelable
import java.text.SimpleDateFormat
import java.util.*

data class Product(
    val name: String = "",
    val expiry: Date = Date(),      //date format: yyyy-mm-dd
    val thumbnail: String = ""
) : Parcelable {

    var id: Long = 0

    private val remainDayBeginWarn = 15

    // 30 day - 15 day - 0day
    // new - nearly expiry - expired
    val state: EXPIRY_STATE
        get() {
            var state = EXPIRY_STATE.NEW
            val now = Calendar.getInstance()
            val elapsedDate = DateUtils.substract(expiry, now.time)
            if (0 < elapsedDate && elapsedDate <= remainDayBeginWarn) {
                state = EXPIRY_STATE.NEARLY_EXPIRY
            } else if (elapsedDate <= 0) {
                state = EXPIRY_STATE.EXPIRED
            }
            return state
        }

    enum class EXPIRY_STATE {
        EXPIRED, NEARLY_EXPIRY, NEW
    }

    init {
        id = ++lastID
    }

    // Parcelable implement
    protected constructor(`in`: Parcel)
            : this(
        name = `in`.readString()!!,
        expiry = SimpleDateFormat(Utils.DATE_PATTERN_VN).parse(`in`.readString()!!)!!,
        thumbnail = `in`.readString()!!
    ) {
        lastID--
        id = `in`.readLong()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(id)
        dest.writeString(name)
        dest.writeString(thumbnail)
        dest.writeString(SimpleDateFormat(Utils.Companion.DATE_PATTERN_VN).format(expiry))
    }

    companion object {
        private var lastID: Long = 99
        val CREATOR: Parcelable.Creator<Product> = object : Parcelable.Creator<Product> {
            override fun createFromParcel(`in`: Parcel): Product? {
                return Product(`in`)
            }

            override fun newArray(size: Int): Array<Product?> {
                return arrayOfNulls(size)
            }
        }
    }
}